ipfsApi/encoding.py

```diff
@@ -42,48 +53,128 @@ class Json(Encoding):
     """
     name = 'json'
 
-    def __init__(self):
-        self.encoder = json.JSONEncoder()
-        self.decoder = json.JSONDecoder()
-
     def parse(self, raw):
-        """Returns a Python object decoded from JSON object(s) in raw.
+        """Returns a Python object decoded from JSON object(s) bytes.
 
-        Some responses from the IPFS api are a concatenated string of JSON
-        objects, which crashes json.loads(), so we need to use this instead as
-        a general approach.
+        Raises
+        ------
+        ~ipfsApi.exceptions.DecodingError
 
         Parameters
         ----------
-        raw : str
+        raw : bytes
             Stringified JSON object
+
+        Returns
+        -------
+            str | list | dict | int
         """
         json_string = raw.strip()
+        decoder = json.JSONDecoder()
         results = []
 
-        obj, idx = self.decoder.raw_decode(json_string)
-        results.append(obj)
-        cur = idx
-        while cur < len(json_string) - 1:
-            if json_string[cur] == '\n':
-                cur += 1
-            obj, idx = self.decoder.raw_decode(json_string[cur:])
+        try:
+            # Python 3 requires this to be a text string
+            if isinstance(json_string, six.binary_type):
+                json_string = json_string.decode("utf-8")
+
+            # Some responses from the IPFS api are a concatenated string of
+            # JSON objects, which crashes json.loads(), so we need to use this
+            # instead as a general approach.
+            obj, idx = decoder.raw_decode(json_string)
             results.append(obj)
-            cur += idx
+            cur = idx
+            while cur < len(json_string) - 1:
+                if json_string[cur] == '\n':
+                    cur += 1
+                obj, idx = decoder.raw_decode(json_string[cur:])
+                results.append(obj)
+                cur += idx
+        except (UnicodeDecodeError, ValueError) as error:
+            raise exceptions.DecodingError('json', error)
 
         if len(results) == 1:
             return results[0]
         return results
 
     def encode(self, obj):
-        """Returns obj serialized as JSON formatted string.
+        """Returns ``obj`` serialized as JSON formatted bytes.
+
+        Raises
+        ------
+        ~ipfsApi.exceptions.EncodingError
 
         Parameters
         ----------
-        obj : str | list | dict | int | None
+        obj : str | list | dict | int
             JSON serializable Python object
+
+        Returns
+        -------
+            bytes
         """
-        return json.dumps(obj)
+        try:
+            result = json.dumps(obj)
+            if isinstance(result, six.text_type):
+                return result.encode("utf-8")
+            else:
+                return result
+        except (UnicodeEncodeError, TypeError) as error:
+            raise exceptions.EncodingError('json', error)
+
+
+class Pickle(Encoding):
```

Comment:
You mean the `Pickle` encoding class. Because it has such a nice symmetry with all the other encoding formats and because it allows one to use `decoder="pickle"` in the client API. :grin: